File: frontend/src/js/components/progressTracker.js
/**
 * Real-time Pipeline Progress Tracker & Background Execution Monitor Component.
 */

import { state } from '../state.js';

export function renderProgressTracker(container) {
  const p = state.pipelineProgress;

  if (!p || p.status === 'IDLE') {
    container.innerHTML = `
      <div class="progress-card progress-idle">
        <div class="progress-header">
          <div class="progress-title">
            <span class="pulse-dot idle"></span>
            <span>Pipeline Engine: Standby</span>
          </div>
          <span class="badge badge-outline">Ready</span>
        </div>
        <p style="font-size: 0.85rem; color: var(--text-muted); margin-top: 0.5rem;">
          Click <strong>Run Ingestion Pipeline</strong> or <strong>Generate Draft Invoices</strong> to process live hotel laundry slips with Gemini Flash OCR.
        </p>
      </div>
    `;
    return;
  }

  const isRunning = p.is_running || p.status === 'RUNNING';
  const isCompleted = p.status === 'COMPLETED';
  const isError = p.status === 'ERROR';

  let statusBadge = '';
  let progressThemeClass = 'progress-running';

  if (isRunning) {
    statusBadge = `<span class="badge badge-primary"><span class="pulse-dot running"></span> Live: Stage ${p.stage_index}/${p.total_stages}</span>`;
    progressThemeClass = 'progress-running';
  } else if (isCompleted) {
    statusBadge = `<span class="badge badge-success">✅ Completed (${p.elapsed_seconds}s)</span>`;
    progressThemeClass = 'progress-completed';
  } else if (isError) {
    statusBadge = `<span class="badge badge-danger">❌ Error</span>`;
    progressThemeClass = 'progress-error';
  }

  const percent = p.percent || 0;
  const stats = p.stats || {};
  const currentStep = p.current_step || 'Processing...';

  container.innerHTML = `
    <div class="progress-card ${progressThemeClass}">
      <div class="progress-header">
        <div class="progress-title">
          <span>${isRunning ? '⚡' : isCompleted ? '🎉' : '⚠️'}</span>
          <span>${p.task_name || 'Daily Pipeline'}</span>
          <span style="font-size: 0.82rem; color: var(--text-muted); font-weight: 500;">
            (${p.month} ${p.year})
          </span>
        </div>
        <div style="display: flex; align-items: center; gap: 0.75rem;">
          ${statusBadge}
          <span style="font-family: var(--font-mono); font-size: 0.95rem; font-weight: 700; color: #ffffff;">
            ${percent}%
          </span>
        </div>
      </div>

      <!-- Animated Glowing Progress Bar -->
      <div class="progress-track">
        <div class="progress-fill ${isRunning ? 'shimmer' : ''}" style="width: ${percent}%;"></div>
      </div>

      <!-- Current Step Description -->
      <div class="progress-step-desc">
        ${isRunning ? '<span class="spinner-sm"></span>' : ''}
        <span>${currentStep}</span>
      </div>

      <!-- Live Real-Time Telemetry Counters -->
      <div class="progress-stats-grid">
        <div class="progress-stat-pill">
          <span class="stat-label">📁 Clients</span>
          <span class="stat-val">${stats.clients_done || 0} / ${stats.clients_total || 0}</span>
        </div>
        <div class="progress-stat-pill">
          <span class="stat-label">🧾 Slips Read</span>
          <span class="stat-val">${stats.slips_processed || 0}</span>
        </div>
        <div class="progress-stat-pill">
          <span class="stat-label">📦 Items Extracted</span>
          <span class="stat-val">${stats.items_extracted || 0}</span>
        </div>
        <div class="progress-stat-pill">
          <span class="stat-label">⚠️ Linen Losses</span>
          <span class="stat-val ${stats.loss_discrepancies > 0 ? 'text-warning' : ''}">${stats.loss_discrepancies || 0} pcs</span>
        </div>
        <div class="progress-stat-pill">
          <span class="stat-label">⏱️ Elapsed Time</span>
          <span class="stat-val font-mono">${p.elapsed_seconds || 0}s</span>
        </div>
      </div>

      ${
        isCompleted && p.last_result?.spreadsheet_url
          ? `
        <div style="margin-top: 1rem; display: flex; justify-content: flex-end;">
          <a href="${p.last_result.spreadsheet_url}" target="_blank" rel="noopener" class="btn btn-outline btn-sm">
            <span>📊</span> Open Updated Google Sheet ↗
          </a>
        </div>
      `
          : ''
      }
    </div>
  `;
}

File: app/utils/progress_tracker.py
"""
Real-time pipeline telemetry and execution progress tracker for ANR Laundry Billing.
Provides live stage updates, percent completion, stats counters, and streaming logs.
"""

from datetime import datetime
import time
from typing import Dict, Any, List, Optional
import threading


class PipelineProgressTracker:
    def __init__(self):
        self._lock = threading.Lock()
        self.is_running: bool = False
        self.status: str = "IDLE"  # "IDLE", "RUNNING", "COMPLETED", "ERROR"
        self.task_name: str = "Idle"
        self.month: str = "August"
        self.year: int = 2026
        self.current_step: str = "System ready."
        self.stage_index: int = 0
        self.total_stages: int = 5
        self.percent: int = 0
        self.stats: Dict[str, Any] = {
            "clients_total": 0,
            "clients_done": 0,
            "slips_processed": 0,
            "items_extracted": 0,
            "loss_discrepancies": 0,
        }
        self.logs: List[Dict[str, Any]] = []
        self.start_timestamp: float = 0.0
        self.started_at: Optional[str] = None
        self.completed_at: Optional[str] = None
        self.elapsed_seconds: float = 0.0
        self.error_message: Optional[str] = None
        self.last_result: Optional[Dict[str, Any]] = None

    def start_pipeline(self, task_name: str, month: str, year: int, total_stages: int = 5):
        with self._lock:
            self.is_running = True
            self.status = "RUNNING"
            self.task_name = task_name
            self.month = month
            self.year = year
            self.current_step = f"Starting {task_name} for {month} {year}..."
            self.stage_index = 1
            self.total_stages = total_stages
            self.percent = 5
            self.stats = {
                "clients_total": 0,
                "clients_done": 0,
                "slips_processed": 0,
                "items_extracted": 0,
                "loss_discrepancies": 0,
            }
            self.start_timestamp = time.time()
            self.started_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            self.completed_at = None
            self.elapsed_seconds = 0.0
            self.error_message = None
            self.last_result = None
            self.logs = []
            
        self.add_log("info", f"🚀 Started {task_name} for {month} {year}.")

    def update_progress(
        self,
        percent: int,
        stage_index: Optional[int] = None,
        current_step: Optional[str] = None,
        stats_update: Optional[Dict[str, Any]] = None,
    ):
        with self._lock:
            self.percent = min(max(percent, 0), 100)
            if stage_index is not None:
                self.stage_index = stage_index
            if current_step is not None:
                self.current_step = current_step
            if stats_update:
                self.stats.update(stats_update)
            if self.start_timestamp > 0:
                self.elapsed_seconds = round(time.time() - self.start_timestamp, 1)

    def add_log(self, level: str, message: str):
        entry = {
            "time": datetime.now().strftime("%H:%M:%S"),
            "level": level,  # "info", "success", "warning", "error"
            "message": message,
        }
        with self._lock:
            self.logs.append(entry)
            if len(self.logs) > 300:
                self.logs.pop(0)

    def complete_pipeline(self, summary: Optional[Dict[str, Any]] = None):
        with self._lock:
            self.is_running = False
            self.status = "COMPLETED"
            self.percent = 100
            self.stage_index = self.total_stages
            self.current_step = "Pipeline completed successfully."
            self.completed_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            if self.start_timestamp > 0:
                self.elapsed_seconds = round(time.time() - self.start_timestamp, 1)
            self.last_result = summary

        slips = self.stats.get("slips_processed", 0)
        items = self.stats.get("items_extracted", 0)
        self.add_log(
            "success",
            f"✅ Completed in {self.elapsed_seconds}s. Processed {slips} slips ({items} line items extracted).",
        )

    def fail_pipeline(self, error_msg: str):
        with self._lock:
            self.is_running = False
            self.status = "ERROR"
            self.current_step = f"Failed: {error_msg}"
            self.error_message = error_msg
            self.completed_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            if self.start_timestamp > 0:
                self.elapsed_seconds = round(time.time() - self.start_timestamp, 1)

        self.add_log("error", f"❌ Pipeline failed: {error_msg}")

    def get_state(self) -> Dict[str, Any]:
        with self._lock:
            if self.is_running and self.start_timestamp > 0:
                self.elapsed_seconds = round(time.time() - self.start_timestamp, 1)
            return {
                "is_running": self.is_running,
                "status": self.status,
                "task_name": self.task_name,
                "month": self.month,
                "year": self.year,
                "current_step": self.current_step,
                "stage_index": self.stage_index,
                "total_stages": self.total_stages,
                "percent": self.percent,
                "stats": dict(self.stats),
                "started_at": self.started_at,
                "completed_at": self.completed_at,
                "elapsed_seconds": self.elapsed_seconds,
                "error_message": self.error_message,
                "recent_logs": list(self.logs[-50:]),
                "last_result": self.last_result,
            }


# Process-wide tracker instance
pipeline_tracker = PipelineProgressTracker()
